File: src/main/java/de/nordgedanken/auto_hot_key/AHKParserDefinition.java
package de.nordgedanken.auto_hot_key;

import com.intellij.lang.ASTNode;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import de.nordgedanken.auto_hot_key.parser.AHKParser;
import de.nordgedanken.auto_hot_key.psi.AHKFile;
import de.nordgedanken.auto_hot_key.psi.AHKTypes;
import org.jetbrains.annotations.NotNull;

import static de.nordgedanken.auto_hot_key.psi.AHKTypes.*;

public class AHKParserDefinition implements ParserDefinition {
    public static final TokenSet WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE);
    public static final TokenSet COMMENTS = TokenSet.create(AHKTypes.COMMENT);
    public static final TokenSet STRING_LITERALS = TokenSet.create(STRING);

    public static final IFileElementType FILE = new IFileElementType(AHKLanguage.INSTANCE);

    @NotNull
    @Override
    public Lexer createLexer(Project project) {
        return new AHKLexerAdapter();
    }

    @NotNull
    @Override
    public TokenSet getWhitespaceTokens() {
        return WHITE_SPACES;
    }

    @NotNull
    @Override
    public TokenSet getCommentTokens() {
        return COMMENTS;
    }

    @NotNull
    @Override
    public TokenSet getStringLiteralElements() {
        return STRING_LITERALS;
    }

    @NotNull
    @Override
    public PsiParser createParser(final Project project) {
        return new AHKParser();
    }

    @Override
    public IFileElementType getFileNodeType() {
        return FILE;
    }

    @Override
    public PsiFile createFile(FileViewProvider viewProvider) {
        return new AHKFile(viewProvider);
    }

    @Override
    public SpaceRequirements spaceExistenceTypeBetweenTokens(ASTNode left, ASTNode right) {
        return SpaceRequirements.MAY;
    }

    @NotNull
    @Override
    public PsiElement createElement(ASTNode node) {
        return AHKTypes.Factory.createElement(node);
    }
}

File: src/main/kotlin/com/autohotkey/lang/parser/AhkParserDefinition.kt
package com.autohotkey.lang.parser

import com.autohotkey.lang.core.AhkLanguage
import com.autohotkey.lang.lexer.AhkLexerAdapter
import com.autohotkey.lang.psi.AhkFile
import com.autohotkey.lang.psi.AhkTypes.Factory
import com.autohotkey.lang.psi.COMMENT_TOKENS
import com.autohotkey.lang.psi.WHITESPACE_TOKENS
import com.intellij.lang.ASTNode
import com.intellij.lang.ParserDefinition
import com.intellij.lang.PsiParser
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.tree.IFileElementType
import com.intellij.psi.tree.TokenSet

class AhkParserDefinition : ParserDefinition {
    override fun createLexer(project: Project?): Lexer = AhkLexerAdapter()

    override fun getWhitespaceTokens() = WHITESPACE_TOKENS

    override fun getCommentTokens() = COMMENT_TOKENS

    override fun getStringLiteralElements(): TokenSet = TokenSet.EMPTY

    override fun createParser(project: Project?): PsiParser = AhkParser()

    override fun getFileNodeType() = IFileElementType(AhkLanguage)

    override fun createFile(viewProvider: FileViewProvider): PsiFile = AhkFile(viewProvider)

    override fun spaceExistenceTypeBetweenTokens(left: ASTNode, right: ASTNode) = ParserDefinition.SpaceRequirements.MAY

    override fun createElement(node: ASTNode?): PsiElement = Factory.createElement(node)
}
